[Logs UI] Log threshold ratio alerts

Kerry350 opened the pull request:
## Summary

This primarily closes #72648, it also closes #73274 and closes #72453. It adds the ability to create ratio based alerts for the log threshold alert type. 

![Screenshot 2020-09-07 at 11 11 52](https://user-images.githubusercontent.com/471693/92383524-66803f00-f106-11ea-99ad-472f4cf36e95.png)

# Implementation notes

- I've stuck with the mathematics principle that dividing by 0 isn't possible, as such if either value is 0 then a ratio is undefined / indeterminate. Given an undefined ratio the alert will **not** fire.

- Ratio alerts are, realistically, what we have already but two sets of criteria compared against each other, as such a lot of the executor logic is reused, but with new ratio results processors and context variables.

- Threshold annotation rendering is turned off for ratio alerts on chart previews.

- I've moved around some folders in preparation for us adding more alert types (e.g. the anomaly alert type). The diff seems to have marked some (not all) of the renames as new files (validation, threshold component etc) so it looks like more code has been added there than in reality. 

# Testing

- Alerting requires SSL (`yarn start --ssl` and `yarn es snapshot --ssl`)

- Quite a lot has been touched, so it would be good to not only verify the new ratio alerts, but also standard count based alerts. Things like UI validation should try to be broken etc.

elasticmachine commented:
Pinging @elastic/logs-metrics-ui (Team:logs-metrics-ui)

kibanamachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/72914/)
* Commit: 829009060deff637d4165c24c388bc3bf0b0acc5
* This comment will update when the build is complete

### Failed CI Steps
* [Check i18n](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/72914/execution/node/370/log)

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/72914/","number":"72914","commit":"829009060deff637d4165c24c388bc3bf0b0acc5"}]}
PIPELINE-->

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/72922/)
* Commit: fbc12838a249969643a02b9b8f1f05816aa40cf3

### Build metrics
#### @kbn/optimizer bundle module count
| id | [value](https://github.com/elastic/kibana/commit/fbc12838a249969643a02b9b8f1f05816aa40cf3) | diff | [baseline](https://github.com/elastic/kibana/commit/ea47ff1f7b6b28862c165709ac9bf93e6aef918f) |
| --- | --- | --- | --- |
| infra | 1094 | +1 | 1093 |

#### async chunks size
| id | [value](https://github.com/elastic/kibana/commit/fbc12838a249969643a02b9b8f1f05816aa40cf3) | diff | [baseline](https://github.com/elastic/kibana/commit/ea47ff1f7b6b28862c165709ac9bf93e6aef918f) |
| --- | --- | --- | --- |
| infra | 3.6MB | +10.3KB | 3.6MB |

#### page load bundle size
| id | [value](https://github.com/elastic/kibana/commit/fbc12838a249969643a02b9b8f1f05816aa40cf3) | diff | [baseline](https://github.com/elastic/kibana/commit/ea47ff1f7b6b28862c165709ac9bf93e6aef918f) |
| --- | --- | --- | --- |
| infra | 279.6KB | +3.0KB | 276.6KB |

### History
* :broken_heart: [Build #72914](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/72914/) failed 829009060deff637d4165c24c388bc3bf0b0acc5

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/72914/","number":"72914","commit":"829009060deff637d4165c24c388bc3bf0b0acc5"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/72922/","number":"72922","commit":"fbc12838a249969643a02b9b8f1f05816aa40cf3"}]}
PIPELINE-->

jasonrhodes reviewed (commented):
I'm taking a look at this PR this week 👍

jasonrhodes reviewed (commented):
Looks good so far! I'm going to take a look at the code more deeply next, but I'm going to add UX based comments individually so far.

jasonrhodes commented:
<img width="284" alt="Screen Shot 2020-09-09 at 1 38 03 PM" src="https://user-images.githubusercontent.com/159370/92641131-48ecda00-f2ac-11ea-9494-b2bb4c2a0fe3.png">

Missing some padding/separation between these sections now, I think.

jasonrhodes commented:
<img width="1552" alt="Screen Shot 2020-09-09 at 1 38 08 PM" src="https://user-images.githubusercontent.com/159370/92641162-586c2300-f2ac-11ea-896d-5d3f97851edd.png">

The loading indicator when opening a new alert is up in the top left corner and not centered (or padded from the edge). Not sure if this is new or not, though.

jasonrhodes commented:
<img width="163" alt="Screen Shot 2020-09-09 at 2 24 57 PM" src="https://user-images.githubusercontent.com/159370/92641243-733e9780-f2ac-11ea-9884-a8ce59b1aba9.png">

I assume this means "more than 75%"? We probably need the % indicator to help with that understanding, both in the form and in the display here. Or use 0.75? Something to help me understand what that number is?

jasonrhodes commented:
<img width="587" alt="Screen Shot 2020-09-09 at 2 25 20 PM" src="https://user-images.githubusercontent.com/159370/92641383-aaad4400-f2ac-11ea-8baf-7f72eaed1cc9.png">

At first I was admittedly pretty lost with what is going on here. It took me a while to realize that when I select "ratio" from the "When" dropdown, I get two separate condition "sets" added and that I can now add and remove conditions to each set which are then evaluated together as a ratio. I'm not sure I have a ton of great ideas to fix this, but as far as a small initial tweak maybe indentation would help? I do think this type of SQL UX may be starting to break down a bit, though, more generally.

@katrin-freihofner do you have thoughts here?

Kerry350 commented:
> Missing some padding/separation between these sections now, I think.

Yeah 👍

> The loading indicator when opening a new alert is up in the top left corner and not centered (or padded from the edge). Not sure if this is new or not, though.

This is old, I *think* this was originally due to the fact we don't really control the initial rendering space, the framework runs our expression code and pops it in there. Things may have changed by now, and positioning that spinner might be easier.

 > I assume this means "more than 75%"? We probably need the % indicator to help with that understanding, both in the form and in the display here. Or use 0.75? Something to help me understand what that number is?

This is the straight up ratio value, as it's the ratio of the count of Case A to the count of Case B. Or CaseA / CaseB. Admittedly 75 isn't a great default, it's what the `count` type uses, something like `2` would be better for twice as many. 

It could be converted to percentages. 

I went with the actual ratio values as that's what other ratio alerting products seemed to use. With ratios I think most people are accustomed to the 3:1  (numerator:denominator) format, and therefore plugging in a real value.

We could make it clear that this is CaseA / CaseB, here is an example from elsewhere:

![Screenshot 2020-09-10 at 12 35 18](https://user-images.githubusercontent.com/471693/92724271-37fea000-f362-11ea-8291-efad1e77f545.png)

> At first I was admittedly pretty lost with what is going on here. It took me a while to realize that when I select "ratio" from the "When" dropdown, I get two separate condition "sets" added and that I can now add and remove conditions to each set which are then evaluated together as a ratio. I'm not sure I have a ton of great ideas to fix this, but as far as a small initial tweak maybe indentation would help? I do think this type of SQL UX may be starting to break down a bit, though, more generally.

I don't have any real ideas here. I did a literal implementation of what was in #72648, which was `When ratio of the count of log entries with alert condition1 AND alert condition2 to the count of log entries with alert condition3 AND alert condition4 is threshold within the last 5 minutes then action.`, when I spoke with @katrin-freihofner it was agreed there'd be no specific design for this, and what was in that ticket was enough.

katrin-freihofner commented:
@jasonrhodes I'm not sure I fully understand the problem. And @Kerry350 did not get any design help as we thought this can move directly to engineering.
From what I see in the screenshots it is what was asked for but I agree, it looks confusing.
I'm wondering if the new expression layout could help here. This is an example:
<img width="579" alt="Screenshot 2020-09-16 at 12 14 19" src="https://user-images.githubusercontent.com/20965076/93324890-c7023100-f816-11ea-824f-41ea458646d4.png">

Also, I think we need to move the chart to the bottom. That said, this is something I would really like to do across Observability and not application by application. Let me know what you are thinking.

jasonrhodes reviewed (commented):
> From what I see in the screenshots it is what was asked for but I agree, it looks confusing.
> I'm wondering if the new expression layout could help here. 

Can we talk about this option at Monday's meeting? This is the kind of solution I'm looking for, for right now -- just something to help orient the user to how these conditions are grouped/organized. We can re-think beyond that after this ticket, if needed.

> Also, I think we need to move the chart to the bottom. That said, this is something I would really like to do across Observability and not application by application. Let me know what you are thinking.

Yeah, this topic has been coming up a lot, and there are valid reasons to keep it where it is but there is also a lot going on with these condition rows. Can we schedule a session at some point next week to talk about Alerting Design overall? I don't think it's urgent but it would be good to start thinking through whether the SQL-style design is going to continue to work or if we need something else, and I think the per-condition previews are one aspect of many there. I'll put something on the calendar and include some folks.

jasonrhodes commented:
> This is old, I _think_ this was originally due to the fact we don't really control the initial rendering space, the framework runs our expression code and pops it in there. Things may have changed by now, and positioning that spinner might be easier.

If it's not super simple, we can address this in another ticket. Not a huge deal.

jasonrhodes commented:
> This is the straight up ratio value, as it's the ratio of the count of Case A to the count of Case B. Or CaseA / CaseB. Admittedly 75 isn't a great default, it's what the `count` type uses, something like `2` would be better for twice as many.
> 
> It could be converted to percentages.
> 
> I went with the actual ratio values as that's what other ratio alerting products seemed to use. With ratios I think most people are accustomed to the 3:1 (numerator:denominator) format, and therefore plugging in a real value.

Ahh ok this makes sense. Changing the default should remove the confusion I had. Sounds good.

Kerry350 commented on x-pack/plugins/infra/public/components/alerting/logs/log_threshold/expression_editor/criteria.tsx:
```diff
@@ -0,0 +1,275 @@
+      {!isRatioAlert(criteria) ? (
+        <CountCriteria {...props} criteria={criteria as CountCriteriaType} errors={errors} />
+      ) : (
+        <RatioCriteria {...props} criteria={criteria as RatioCriteriaType} errors={errors} />
```
I struggled to get TS to infer things correctly 😞 

I tried to keep the casting limited to top levels, but I agree I don't want the casting there at all. 

I'm going to revisit this now and see if I can get an agreement between me and TypeScript...

(There is one way I know it'll work but it comes with a io-ts decode overhead so I didn't want to use that).

Kerry350 replied:
Okay, found out how to do this properly. To pass on the knowledge:

I added a [User-defined type guard](https://www.typescriptlang.org/docs/handbook/advanced-types.html#user-defined-type-guards).

Which looks like so:

```
export const isRatioAlert = (criteria: AlertParams['criteria']): criteria is RatioCriteria => {
  return criteria.length > 0 && Array.isArray(criteria[0]) ? true : false;
};
```

The key being the `is` portion.

jasonrhodes replied:
Oh yes I've seen this before! Great work!!

Kerry350 commented:
@jasonrhodes This is ready for another look, I've addressed all of the feedback here. Recap on the main bits:

- New design is in place

![Screenshot 2020-09-24 at 14 52 48](https://user-images.githubusercontent.com/471693/94154641-06172e80-fe76-11ea-91a3-77baec0fc201.png)

- Used a default of `2` vs `75` for ratio alert thresholds
- Removed all type casting in favour of user-defined type guards

A note on the loading spinner: this is actually the spinner from the alerting framework, and not something we control. The spinner we control is this one (which is when we're loading the fields information, if needed):

![Screenshot 2020-09-24 at 12 23 44](https://user-images.githubusercontent.com/471693/94155067-71f99700-fe76-11ea-94ce-5310dbdc4c74.png)

So I haven't changed anything there.

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/76928/)
* Commit: cdcc4b133af18a0d61c309cba12c283642eb58a0

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### @kbn/optimizer bundle module count
| id | [value](https://github.com/elastic/kibana/commit/cdcc4b133af18a0d61c309cba12c283642eb58a0) | diff | [baseline](https://github.com/elastic/kibana/commit/ba635340bee7535a53f441ec70b40a81d11d0f44) |
| --- | --- | --- | --- |
| infra | 1125 | +1 | 1124 |

#### async chunks size
| id | [value](https://github.com/elastic/kibana/commit/cdcc4b133af18a0d61c309cba12c283642eb58a0) | diff | [baseline](https://github.com/elastic/kibana/commit/ba635340bee7535a53f441ec70b40a81d11d0f44) |
| --- | --- | --- | --- |
| infra | 4.2MB | +11.2KB | 4.2MB |

#### page load bundle size
| id | [value](https://github.com/elastic/kibana/commit/cdcc4b133af18a0d61c309cba12c283642eb58a0) | diff | [baseline](https://github.com/elastic/kibana/commit/ba635340bee7535a53f441ec70b40a81d11d0f44) |
| --- | --- | --- | --- |
| infra | 280.5KB | +3.1KB | 277.3KB |

### History
* :green_heart: [Build #72922](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/72922/) succeeded fbc12838a249969643a02b9b8f1f05816aa40cf3
* :broken_heart: [Build #72914](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/72914/) failed 829009060deff637d4165c24c388bc3bf0b0acc5

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/72914/","number":"72914","commit":"829009060deff637d4165c24c388bc3bf0b0acc5"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/72922/","number":"72922","commit":"fbc12838a249969643a02b9b8f1f05816aa40cf3"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/76928/","number":"76928","commit":"cdcc4b133af18a0d61c309cba12c283642eb58a0"}]}
PIPELINE-->

Kerry350 commented:
@elasticmachine merge upstream

kibanamachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/77440/)
* Commit: e46a9c748ce16e42586c7039e095f2d7f3a1ae85
* This comment will update when the build is complete

### Failed CI Steps
* [Check Types](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/77440/execution/node/374/log)

### History
* :green_heart: [Build #76928](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/76928/) succeeded cdcc4b133af18a0d61c309cba12c283642eb58a0
* :green_heart: [Build #72922](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/72922/) succeeded fbc12838a249969643a02b9b8f1f05816aa40cf3
* :broken_heart: [Build #72914](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/72914/) failed 829009060deff637d4165c24c388bc3bf0b0acc5

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/72914/","number":"72914","commit":"829009060deff637d4165c24c388bc3bf0b0acc5"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/72922/","number":"72922","commit":"fbc12838a249969643a02b9b8f1f05816aa40cf3"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/76928/","number":"76928","commit":"cdcc4b133af18a0d61c309cba12c283642eb58a0"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/77440/","number":"77440","commit":"e46a9c748ce16e42586c7039e095f2d7f3a1ae85"}]}
PIPELINE-->

Kerry350 commented:
@elasticmachine merge upstream

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/77794/)
* Commit: e97686f84d39466f18830235acdd490df6644b23

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### @kbn/optimizer bundle module count
| id | [value](https://github.com/elastic/kibana/commit/e97686f84d39466f18830235acdd490df6644b23) | diff | [baseline](https://github.com/elastic/kibana/commit/61aac716b3f0d3483fb53ad3f64f9356b304a59d) |
| --- | --- | --- | --- |
| `infra` | 1126 | +1 | 1125 |

#### async chunks size
| id | [value](https://github.com/elastic/kibana/commit/e97686f84d39466f18830235acdd490df6644b23) | diff | [baseline](https://github.com/elastic/kibana/commit/61aac716b3f0d3483fb53ad3f64f9356b304a59d) |
| --- | --- | --- | --- |
| `infra` | 4.2MB | +11.3KB | 4.2MB |

#### page load bundle size
| id | [value](https://github.com/elastic/kibana/commit/e97686f84d39466f18830235acdd490df6644b23) | diff | [baseline](https://github.com/elastic/kibana/commit/61aac716b3f0d3483fb53ad3f64f9356b304a59d) |
| --- | --- | --- | --- |
| `infra` | 137.7KB | +3.2KB | 134.5KB |

### History
* :broken_heart: [Build #77440](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/77440/) failed e46a9c748ce16e42586c7039e095f2d7f3a1ae85
* :green_heart: [Build #76928](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/76928/) succeeded cdcc4b133af18a0d61c309cba12c283642eb58a0
* :green_heart: [Build #72922](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/72922/) succeeded fbc12838a249969643a02b9b8f1f05816aa40cf3
* :broken_heart: [Build #72914](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/72914/) failed 829009060deff637d4165c24c388bc3bf0b0acc5

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/72914/","number":"72914","commit":"829009060deff637d4165c24c388bc3bf0b0acc5"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/72922/","number":"72922","commit":"fbc12838a249969643a02b9b8f1f05816aa40cf3"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/76928/","number":"76928","commit":"cdcc4b133af18a0d61c309cba12c283642eb58a0"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/77440/","number":"77440","commit":"e46a9c748ce16e42586c7039e095f2d7f3a1ae85"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/77794/","number":"77794","commit":"e97686f84d39466f18830235acdd490df6644b23"}]}
PIPELINE-->

jasonrhodes reviewed (approved):
LGTM @Kerry350 ! Thanks so much for working with us on getting this ready to merge!
